### ISSUE TITLE
Stellar Core crashes after downloading a bad file from history

### ISSUE THREAD

**Issue by oryband** [stellar/stellar-core#1675]
# Issue type
Bug

# Version found
v9.2.0

# Your environment and setup
  * Ubuntu 16.04
  * stellar-core is dockerized
  * Docker 17.12.1-ce
  * Compose 1.18.0, build 8dd22a9

# Issue description
After finishing a complete catchup for a certain stellar-core node using `CATCHUP_COMPLETE=true`, core suddenly loops back to catching up again from the start (on 0% when it just reached 100%), then crashes with the following error:
```
[History INFO] Catching up: verifying checkpoint 1/10385 (0%)
[Fs ERROR] failed to open XDR file: buckets/tmp/catchup-52d5d909f4b62698/ledger-0000003f.xdr, reason: 2 [XDRStream.h:49]
terminate called after throwing an instance of 'std::runtime_error'
   what():  failed to open XDR file: buckets/tmp/catchup-52d5d909f4b62698/ledger-0000003f.xdr, reason: 2
```

## Expected result
stellar-core should not repeat catching up after reaching 100%. That core history archive is already complete.

### Combined log output (stellar-core, dmesg, syslog):

```
Finishing previous catchup
--------------------------

2018-04-27T00:08:16.436 GBT6G [History INFO] Catching up: downloading ledger files 10381/10385 (99%)
Completed 5.2 KiB/5.2 KiB (3.3 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-ap-southeast-1/ledger/00/0a/1f/ledger-000a1f3f.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a1f3f.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (4.1 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-eu-central-1/ledger/00/0a/1f/ledger-000a1f7f.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a1f7f.xdr.gz.tmp
2018-04-27T00:08:16.720 GBT6G [History INFO] Catching up: downloading ledger files 10382/10385 (99%)
2018-04-27T00:08:16.744 GBT6G [History INFO] Catching up: downloading ledger files 10383/10385 (99%)
Completed 5.2 KiB/5.2 KiB (4.0 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-eu-central-1/ledger/00/0a/20/ledger-000a20bf.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a20bf.xdr.gz.tmp
2018-04-27T00:08:17.819 GBT6G [History INFO] Catching up: downloading ledger files 10384/10385 (99%)
Completed 5.2 KiB/5.2 KiB (15.9 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-us-east-1/ledger/00/0a/22/ledger-000a227f.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a227f.xdr.gz.tmp
2018-04-27T00:08:18.094 GBT6G [History INFO] Catching up: downloading ledger files 10385/10385 (100%)
Completed 5.2 KiB/5.2 KiB (3.0 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-sa-east-1/ledger/00/0a/1f/ledger-000a1fbf.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a1fbf.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (15.7 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-us-east-1/ledger/00/0a/22/ledger-000a22bf.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a22bf.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (3.4 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-ap-southeast-1/ledger/00/0a/20/ledger-000a207f.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a207f.xdr.gz.tmp
Completed 5.3 KiB/5.3 KiB (3.3 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-sa-east-1/ledger/00/0a/20/ledger-000a203f.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a203f.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (17.6 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-us-east-1/ledger/00/0a/23/ledger-000a233f.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a233f.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (3.2 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-ap-southeast-1/ledger/00/0a/1f/ledger-000a1fff.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a1fff.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (15.7 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-us-east-1/ledger/00/0a/23/ledger-000a237f.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a237f.xdr.gz.tmp
2018-04-27T00:08:18.724 GBT6G [Herder INFO] Quorum information for 666043 : {"agree":5,"disagree":0,"fail_at":2,"hash":"4b9681","missing":0,"phase":"EXTERNALIZE"}
2018-04-27T00:08:18.727 GBT6G [Ledger INFO] Got consensus: [seq=666045, prev=4b0f6b, tx_count=0, sv: [  txH: 736416, ct: 1524787697, upgrades: [ ] ]]
Completed 5.2 KiB/5.2 KiB (4.1 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-eu-central-1/ledger/00/0a/21/ledger-000a21ff.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a21ff.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (4.2 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-eu-central-1/ledger/00/0a/22/ledger-000a223f.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a223f.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (3.4 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-ap-southeast-1/ledger/00/0a/20/ledger-000a20ff.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a20ff.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (15.4 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-us-east-1/ledger/00/0a/23/ledger-000a23ff.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a23ff.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (3.3 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-sa-east-1/ledger/00/0a/21/ledger-000a213f.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a213f.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (4.3 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-eu-central-1/ledger/00/0a/22/ledger-000a22ff.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a22ff.xdr.gz.tmp
Completed 5.2 KiB/5.2 KiB (4.2 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-eu-central-1/ledger/00/0a/23/ledger-000a23bf.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a23bf.xdr.gz.tmp
2018-04-27T00:08:23.559 GBT6G [Herder INFO] Quorum information for 666044 : {"agree":5,"disagree":0,"fail_at":2,"hash":"4b9681","missing":0,"phase":"EXTERNALIZE"}
2018-04-27T00:08:23.561 GBT6G [Ledger INFO] Got consensus: [seq=666046, prev=f023bf, tx_count=0, sv: [  txH: db2b8a, ct: 1524787702, upgrades: [ ] ]]
Completed 5.2 KiB/5.2 KiB (3.3 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-ap-southeast-1/ledger/00/0a/24/ledger-000a243f.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-000a243f.xdr.gz.tmp
2018-04-27T00:08:28.559 GBT6G [Herder INFO] Quorum information for 666045 : {"agree":5,"disagree":0,"fail_at":2,"hash":"4b9681","missing":0,"phase":"EXTERNALIZE"}
2018-04-27T00:08:28.561 GBT6G [Ledger INFO] Got consensus: [seq=666047, prev=ffdc71, tx_count=0, sv: [  txH: d46826, ct: 1524787707, upgrades: [ ] ]]
2018-04-27T00:08:34.598 GBT6G [Herder INFO] Quorum information for 666046 : {"agree":5,"disagree":0,"fail_at":2,"hash":"4b9681","missing":0,"phase":"EXTERNALIZE"}
2018-04-27T00:08:34.600 GBT6G [Ledger INFO] Got consensus: [seq=666048, prev=ee0629, tx_count=0, sv: [  txH: 80067d, ct: 1524787712, upgrades: [ ] ]]
2018-04-27T00:08:38.719 GBT6G [Herder INFO] Quorum information for 666047 : {"agree":5,"disagree":0,"fail_at":2,"hash":"4b9681","missing":0,"phase":"EXTERNALIZE"}
2018-04-27T00:08:38.721 GBT6G [Ledger INFO] Got consensus: [seq=666049, prev=2b8aec, tx_count=0, sv: [  txH: 233dc6, ct: 1524787717, upgrades: [ ] ]]
2018-04-27T00:08:46.840 GBT6G [Herder INFO] Quorum information for 666048 : {"agree":5,"disagree":0,"fail_at":2,"hash":"4b9681","missing":0,"phase":"EXTERNALIZE"}
2018-04-27T00:08:46.843 GBT6G [Ledger INFO] Got consensus: [seq=666050, prev=f2f7df, tx_count=0, sv: [  txH: 6ec378, ct: 1524787722, upgrades: [ ] ]]
2018-04-27T00:08:48.557 GBT6G [Herder INFO] Quorum information for 666049 : {"agree":5,"disagree":0,"fail_at":2,"hash":"4b9681","missing":0,"phase":"EXTERNALIZE"}
2018-04-27T00:08:48.559 GBT6G [Ledger INFO] Got consensus: [seq=666051, prev=0a6acf, tx_count=0, sv: [  txH: 68e57e, ct: 1524787727, upgrades: [ ] ]]
2018-04-27T00:08:56.965 GBT6G [Herder INFO] Quorum information for 666050 : {"agree":5,"disagree":0,"fail_at":2,"hash":"4b9681","missing":0,"phase":"EXTERNALIZE"}

A while later...
-----------------

2018-04-27T08:48:14.634 GBT6G [Ledger INFO] Got consensus: [seq=672257, prev=276c2a, tx_count=0, sv: [  txH: b40b86, ct: 1524818893, upgrades: [ ] ]]
2018-04-27T08:48:19.631 GBT6G [Herder INFO] Quorum information for 672256 : {"agree":5,"disagree":0,"fail_at":2,"hash":"4b9681","missing":0,"phase":"EXTERNALIZE"}
2018-04-27T08:48:19.634 GBT6G [Ledger INFO] Got consensus: [seq=672258, prev=4b2847, tx_count=0, sv: [  txH: a17804, ct: 1524818898, upgrades: [ ] ]]
2018-04-27T08:48:24.631 GBT6G [Herder INFO] Quorum information for 672257 : {"agree":5,"disagree":0,"fail_at":2,"hash":"4b9681","missing":0,"phase":"EXTERNALIZE"}
2018-04-27T08:48:24.634 GBT6G [Ledger INFO] Got consensus: [seq=672259, prev=c96b82, tx_count=0, sv: [  txH: 2024ea, ct: 1524818903, upgrades: [ ] ]]
Completed 5.2 KiB/5.2 KiB (15.9 KiB/s) with 1 file(s) remaining download: s3://stellar-core-my-testnet-us-east-1/ledger/00/01/b7/ledger-0001b77f.xdr.gz to buckets/tmp/catchup-52d5d909f4b62698/ledger-0001b77f.xdr.gz.tmp
2018-04-27T08:48:27.385 GBT6G [History INFO] Catchup verifying ledger chain for checkpointRange [1..664622]
2018-04-27T08:48:27.385 GBT6G [History INFO] Catching up: verifying checkpoint 1/10385 (0%)
2018-04-27T08:48:27.385 GBT6G [Fs ERROR] failed to open XDR file: buckets/tmp/catchup-52d5d909f4b62698/ledger-0000003f.xdr, reason: 2 [XDRStream.h:49]
terminate called after throwing an instance of 'std::runtime_error'
  what():  failed to open XDR file: buckets/tmp/catchup-52d5d909f4b62698/ledger-0000003f.xdr, reason: 2
2018-04-27T08:48:27.971 GBT6G [default INFO] Starting stellar-core v9.2.0
2018-04-27T08:48:27.971 GBT6G [default INFO] Config from stellar-core.cfg
2018-04-27T08:48:27.971 GBT6G [Database INFO] Connecting to: postgresql://dbname=core host=stellar-core-my-testnet-us-west-1.c3ofnhfeeiwz.us-west-1.rds.amazonaws.com user=stellar password=********
2018-04-27T08:48:28.014 GBT6G [SCP INFO] LocalNode::LocalNode@GBT6G qSet: 4b9681
2018-04-27T08:48:28.014 GBT6G [default INFO] Listening on 0.0.0.0:11626 for HTTP requests
2018-04-27T08:48:28.015 GBT6G [History INFO] Archive 'cache' has 'put' and 'get' commands, will be read and written
2018-04-27T08:48:28.015 GBT6G [History INFO] Archive 's3' has 'put' and 'get' commands, will be read and written
2018-04-27T08:48:28.015 GBT6G [History INFO] Archive 's3-ap-southeast-1' has 'get' command only, will not be written
2018-04-27T08:48:28.015 GBT6G [History INFO] Archive 's3-eu-central-1' has 'get' command only, will not be written
2018-04-27T08:48:28.015 GBT6G [History INFO] Archive 's3-sa-east-1' has 'get' command only, will not be written
2018-04-27T08:48:28.015 GBT6G [History INFO] Archive 's3-us-east-1' has 'get' command only, will not be written
2018-04-27T08:48:28.016 GBT6G [default INFO] ll?level=info -> {
   "Global" : "Info"
}
2018-04-27T08:48:28.017 GBT6G [default INFO] Loading last known ledger
2018-04-27T08:48:28.019 GBT6G [Ledger INFO] Loaded last known ledger: [seq=1, hash=0d0742]
2018-04-27T08:48:28.585 GBT6G [Overlay INFO] New connected peer (empty)
2018-04-27T08:48:28.713 GBT6G [Overlay INFO] New connected peer (empty)
2018-04-27T08:48:28.770 GBT6G [Overlay INFO] successful handshake with sa-east-1@54.94.132.10:11625
2018-04-27T08:48:28.897 GBT6G [Overlay INFO] successful handshake with ap-southeast-1@52.77.237.143:11625
2018-04-27T08:48:29.028 GBT6G [Overlay INFO] New connected peer (empty)
2018-04-27T08:48:29.095 GBT6G [Overlay INFO] successful handshake with us-east-1@35.173.181.102:11625
2018-04-27T08:48:30.211 GBT6G [Overlay INFO] New connected peer 18.196.164.119:11625
2018-04-27T08:48:30.664 GBT6G [Overlay INFO] successful handshake with eu-central-1@18.196.164.119:11625
2018-04-27T08:49:03.019 GBT6G [Herder INFO] Lost track of consensus
2018-04-27T08:49:03.019 GBT6G [Herder INFO] Out of sync context: {
   "queue" : null,
   "you" : "us-west-1"
}
2018-04-27T08:49:04.635 GBT6G [Ledger INFO] Got consensus: [seq=672267, prev=53dd75, tx_count=0, sv: [  txH: 6b7acd, ct: 1524818943, upgrades: [ ] ]]
2018-04-27T08:49:04.635 GBT6G [Ledger INFO] Lost sync, local LCL is 1, network closed ledger 672267
2018-04-27T08:49:04.635 GBT6G [Ledger INFO] Close of ledger 672267 buffered, starting catchup
2018-04-27T08:49:04.635 GBT6G [Ledger INFO] Changing state LM_BOOTING_STATE -> LM_CATCHING_UP_STATE
2018-04-27T08:49:04.636 GBT6G [History INFO] Starting catchup with configuration:
  lastClosedLedger: 1
  toLedger: 672266
  count: 4294967295
2018-04-27T08:49:04.636 GBT6G [History INFO] Catchup downloading history archive state at checkpoint 672319
2018-04-27T08:49:04.636 GBT6G [Work WARNING] Scheduling retry #1/5 in 36 sec, for get-history-archive-state
2018-04-27T08:49:09.632 GBT6G [Ledger INFO] Got consensus: [seq=672268, prev=774aa0, tx_count=0, sv: [  txH: 738bba, ct: 1524818948, upgrades: [ ] ]]
2018-04-27T08:49:09.632 GBT6G [History INFO] Catching up: Awaiting checkpoint (ETA: 31 seconds)
2018-04-27T08:49:14.630 GBT6G [Herder INFO] Quorum information for 672267 : {"agree":5,"disagree":0,"fail_at":2,"hash":"4b9681","missing":0,"phase":"EXTERNALIZE"}

At the same time:

kern.log (dmesg)
-----------

Apr 27 08:48:27 ip-172-31-19-59 kernel: [1941476.694974] traps: stellar-core[15656] general protection ip:7f0bb0bb7196 sp:7ffef3c3a2e0 error:0 in libc-2.23.so[7f0bb0b80000+1c0000]

syslog
------

Apr 27 08:48:27 ip-172-31-19-59 kernel: [1941476.694974] traps: stellar-core[15656] general protection ip:7f0bb0bb7196 sp:7ffef3c3a2e0 error:0 in libc-2.23.so[7f0bb0b80000+1c0000]
Apr 27 08:48:27 ip-172-31-19-59 dockerd[1118]: time="2018-04-27T08:48:27Z" level=info msg="shim reaped" id=4993c1bee804ac7ebf200589d2f30b459b579c3bc2e61274f8b8f5161e62020f module="containerd/tasks"
Apr 27 08:48:27 ip-172-31-19-59 dockerd[1118]: time="2018-04-27T08:48:27.566517269Z" level=info msg="ignoring event" module=libcontainerd namespace=moby topic=/tasks/delete type="*events.TaskDelete"
```

stellar-core.cfg:
```toml
HTTP_PORT=11626
PUBLIC_HTTP_PORT=true

LOG_FILE_PATH=""

COMMANDS=["ll?level=info"]

DATABASE="postgresql://dbname=core host=xxx user=stellar password=xxx"

NETWORK_PASSPHRASE="my passphrase"

CATCHUP_COMPLETE=true

NODE_SEED="xxx us-west-1"

NODE_IS_VALIDATOR=true

TARGET_PEER_CONNECTIONS=4

KNOWN_PEERS=[
 "stellar-core-us-east-1.my-testnet.com:11625",
 "stellar-core-eu-central-1.my-testnet.com:11625",
 "stellar-core-sa-east-1.my-testnet.com:11625",
 "stellar-core-ap-southeast-1.my-testnet.com:11625",
]

NODE_NAMES=[
 "xxx1  us-east-1",
 "xxx2  eu-central-1",
 "xxx3  sa-east-1",
 "xxx4  ap-southeast-1",
]

[QUORUM_SET]
THRESHOLD_PERCENT=66
VALIDATORS=[
 "$us-east-1",
 "$us-west-1",
 "$eu-central-1",
 "$sa-east-1",
 "$ap-southeast-1",
]

[HISTORY.cache]
get="cp /tmp/stellar-core/history/vs/{0} {1}"
put="cp {0} /tmp/stellar-core/history/vs/{1}"
mkdir="mkdir -p /tmp/stellar-core/history/vs/{0}"

[HISTORY.s3]
get="aws s3 cp s3://stellar-core-my-testnet-us-west-1/{0} {1}"
put="aws s3 cp {0} s3://stellar-core-my-testnet-us-west-1/{1}"

[HISTORY.s3-us-east-1]
get="aws s3 cp s3://stellar-core-my-testnet-us-east-1/{0} {1}"

[HISTORY.s3-eu-central-1]
get="aws s3 cp s3://stellar-core-my-testnet-eu-central-1/{0} {1}"

[HISTORY.s3-sa-east-1]
get="aws s3 cp s3://stellar-core-my-testnet-sa-east-1/{0} {1}"

[HISTORY.s3-ap-southeast-1]
get="aws s3 cp s3://stellar-core-my-testnet-ap-southeast-1/{0} {1}"


# vi: ft=toml
```

Dockerfile:
```Dockerfile
FROM ubuntu:16.04

ENV STELLAR_CORE_VERSION v9.2.0

RUN BUILD_DEPS="git build-essential pkg-config autoconf automake libtool bison flex libpq-dev clang++-3.5 gcc-4.9 g++-4.9 cpp-4.9 pandoc"; \
    RUN_DEPS="libpq5"; \
    apt-get -qq update && apt-get -qq install $BUILD_DEPS $RUN_DEPS \
    && git clone -q --branch ${STELLAR_CORE_VERSION} --recurse-submodules https://github.com/stellar/stellar-core.git /tmp/stellar-core \
    && cd /tmp/stellar-core \
    && ./autogen.sh \
    && ./configure \
    && make -j `cat /proc/cpuinfo | awk '/^processor/{print $3}' | tail -1` \
    && make check \
    && make install \
    && make clean \
    && cd - && rm -rf /tmp/stellar-core \
    && apt-get -qq autoremove $BUILD_DEPS \
    && rm -rf /var/lib/apt/lists/*

# (optional) install aws cli
# required for publishing history to s3
RUN apt-get -qq update && apt-get -qq install python3-pip \
    && pip3 -qq install awscli \
    && rm -rf /var/lib/apt/lists/*

RUN mkdir -p \
    /opt/stellar-core \
    /tmp/stellar-core

VOLUME ["/opt/stellar-core", "/tmp/stellar-core"]

WORKDIR /opt/stellar-core

EXPOSE 11625 11626

ENTRYPOINT ["/usr/local/bin/stellar-core"]
```
stellar-core container setup (`docker-compose.yml`):

```yaml
version: "3"

services:
  stellar-core:
    image: my-org/stellar-core
    build:
      dockerfile: Dockerfile.stellar-core
      context: .
    ports:
      - 11625:11625
      - 11626:11626
    volumes:
      - ./stellar-core/opt/stellar-core:/opt/stellar-core
      - ./stellar-core/tmp/stellar-core:/tmp/stellar-core
    restart: on-failure
    logging:
      driver: json-file
      options:
        max-size: 100m
        max-file: "3"
```

**Comment by MonsieurNicolas** [comment]
(update path to the ledger xdr.gz file, it should be correct now)

**Comment by MonsieurNicolas** [comment]
@oryband , did you get a chance to look at the logs?

**Comment by oryband** [comment]
@MonsieurNicolas no. Sorry but I don't have access to the logs anymore.

**Comment by MonsieurNicolas** [comment]
I think this will be fixed when the new `Work` interface lands

**Comment by MonsieurNicolas** [comment]
@marta-lokhova is this fixed now?

**Comment by marta-lokhova** [comment]
I think this is a duplicate of #1676, which has been fixed in #2078
